std/visibility-check/src/main/java/br/com/orcinus/orca/std/visibility/check/detection/PackageProtectedDetector.kt

```diff
@@ -0,0 +1,98 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See
+ * the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program. If
+ * not, see https://www.gnu.org/licenses.
+ */
+
+package br.com.orcinus.orca.std.visibility.check.detection
+
+import br.com.orcinus.orca.std.visibility.PackageProtected
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Incident
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import org.jetbrains.uast.UCallExpression
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UExpression
+import org.jetbrains.uast.getQualifiedChain
+
+/**
+ * [Detector] that reports accesses to structures marked as package-protected that have been made
+ * from a package that is neither the one in which they have been declared nor one of its children
+ * (i. e., a class marked as package-protected declared at `br.com.orcinus.orca.core` can be
+ * accessed from `br.com.orcinus.orca.core` and `br.com.orcinus.orca.core.sample`, but shouldn't be
+ * accessed from `br.com.orcinus.orca.app`).
+ *
+ * A structure is considered to have package-protected visibility if it's annotated with
+ * [PackageProtected].
+ */
+internal class PackageProtectedDetector : Detector(), SourceCodeScanner {
+  override fun getApplicableUastTypes(): List<Class<out UElement>> {
+    return listOf(UCallExpression::class.java)
+  }
+
+  override fun createUastHandler(context: JavaContext): UElementHandler {
+    return object : UElementHandler() {
+      override fun visitCallExpression(node: UCallExpression) {
+        reportExpressionsResolvedToDeclarationsMarkedAsPackageProtectedReferencedFromOutsidePackage(
+          context,
+          node
+        )
+      }
+    }
+  }
+
+  /**
+   * Reports references to package-protected structures from outside the package in which they have
+   * been declared or its children in each of the qualified [UExpression]s within the [expression].
+   *
+   * @param context [JavaContext] with which [UExpression] that refer to package-private structures
+   *   from an outside package will be obtained from the [expression].
+   * @param expression [UExpression] to be checked.
+   */
+  private fun reportExpressionsResolvedToDeclarationsMarkedAsPackageProtectedReferencedFromOutsidePackage(
+    context: JavaContext,
+    expression: UExpression
+  ) {
+    Incident(context, issue).let { incident ->
```

Comment:
An incident is unnecessarily being created for *each* call expression that is analyzed.